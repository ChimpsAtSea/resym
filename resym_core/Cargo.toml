[package]
name = "resym_core"
version.workspace = true
edition = "2021"
rust-version = "1.75"
exclude = [
    "tests/data/*",
    "tests/snapshots/*",
]

[features]
default = ["rayon", "http"]

rayon = ["dep:rayon", "dep:num_cpus", "dashmap/rayon"]
http = ["dep:ehttp", "dep:url"]

[dependencies]
pdb = { git = "https://github.com/ergrelet/pdb", branch = "0.8.0-resym" }
thiserror = "1.0"
log = "0.4"
rayon = { version = "1.5", optional = true }
num_cpus = { version = "1.13", optional = true }
crossbeam-channel = "0.5"
dashmap = { version = "5.2" }
regex = "1.5"
similar = "2.1"
ehttp = { version = "0.2", optional = true }
url = { version = "2.3", optional = true }

# Web:
[target.'cfg(target_arch = "wasm32")'.dependencies]
wasm_thread = { version = "0.2", features = ["es_modules"] }
instant = { version = "0.1", features = ["wasm-bindgen"] }

[dev-dependencies]
insta = "1.14"
