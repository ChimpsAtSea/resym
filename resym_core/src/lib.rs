pub mod backend;
pub mod diffing;
mod error;
pub mod frontend;
pub mod pdb_file;
pub mod pdb_types;
pub mod syntax_highlighting;

pub use error::*;

const PKG_VERSION: &str = env!("CARGO_PKG_VERSION");

/// Macro used to switch between iterators depending on rayon's availability
#[macro_export]
#[cfg(not(feature = "rayon"))]
macro_rules! cond_par_iter {
    ($expression:expr) => {
        $expression.iter()
    };
}
#[macro_export]
#[cfg(feature = "rayon")]
macro_rules! cond_par_iter {
    ($expression:expr) => {
        $expression.par_iter()
    };
}

/// Macro used to switch between functions depending on rayon's availability
#[macro_export]
#[cfg(not(feature = "rayon"))]
macro_rules! cond_sort_by {
    ($expression:expr, $($x:tt)*) => {
        $expression.sort_by($($x)*)
    };
}
#[macro_export]
#[cfg(feature = "rayon")]
macro_rules! cond_sort_by {
    ($expression:expr, $($x:tt)*) => {
        $expression.par_sort_by($($x)*)
    };
}
