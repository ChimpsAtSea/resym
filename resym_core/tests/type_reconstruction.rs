use std::path::Path;

use resym_core::{
    pdb_file::PdbFile, pdb_types::AccessSpecifierReconstructionFlavor,
    pdb_types::PrimitiveReconstructionFlavor,
};

const TEST_PDB_FILE_PATH: &str = "tests/data/test.pdb";
const TEST_CASES: &[&str] = &[
    "resym_test::PrimitiveTypesTest",
    "resym_test::ArrayTest",
    "resym_test::BitFieldsTest1",
    "resym_test::BitFieldsTest2",
    "resym_test::UnionTest",
    "resym_test::StructTest",
    "resym_test::EnumTest1",
    "resym_test::EnumTest2",
    "resym_test::StructUnnamedUdtTest1",
    "resym_test::StructUnnamedUdtTest2",
    "resym_test::StructUnnamedUdtTest3",
    "resym_test::UnionUnnamedUdtTest1",
    "resym_test::PureVirtualClassSpecialized",
    "resym_test::InterfaceImplClass",
    "resym_test::SpecializedInterfaceImplClass",
    "resym_test::ClassWithRefsAndStaticsTest",
    "resym_test::BigOffsetsStruct",
    "resym_test::BitFieldsTest3",
    "resym_test::BitFieldsTest4",
    "resym_test::BitFieldsTest5",
    "resym_test::BitFieldsTest6",
    "resym_test::BitFieldsTest7",
    "resym_test::NestedStructUnionRegression1",
    "resym_test::NtdllRegression1",
];

#[test]
fn test_type_reconstruction_portable_access_specifiers() {
    test_type_reconstruction_internal(
        "type_reconstruction_portable_access_specifiers",
        PrimitiveReconstructionFlavor::Portable,
        AccessSpecifierReconstructionFlavor::Always,
        false,
        true,
        true,
        true,
        false,
        false,
    );
}

#[test]
fn test_type_reconstruction_microsoft_access_specifiers() {
    test_type_reconstruction_internal(
        "type_reconstruction_microsoft_access_specifiers",
        PrimitiveReconstructionFlavor::Microsoft,
        AccessSpecifierReconstructionFlavor::Always,
        false,
        true,
        true,
        true,
        false,
        false,
    );
}

#[test]
fn test_type_reconstruction_raw_access_specifiers() {
    test_type_reconstruction_internal(
        "type_reconstruction_raw_access_specifiers",
        PrimitiveReconstructionFlavor::Raw,
        AccessSpecifierReconstructionFlavor::Always,
        false,
        true,
        true,
        true,
        false,
        false,
    );
}

#[test]
fn test_type_reconstruction_msvc_access_specifiers() {
    test_type_reconstruction_internal(
        "type_reconstruction_msvc_access_specifiers",
        PrimitiveReconstructionFlavor::Msvc,
        AccessSpecifierReconstructionFlavor::Always,
        false,
        true,
        true,
        true,
        false,
        false,
    );
}

#[test]
fn test_type_reconstruction_automatic_access_specifiers() {
    test_type_reconstruction_internal(
        "type_reconstruction_automatic_access_specifiers",
        PrimitiveReconstructionFlavor::Portable,
        AccessSpecifierReconstructionFlavor::Automatic,
        false,
        true,
        true,
        true,
        false,
        false,
    );
}

#[test]
fn test_type_reconstruction_no_size_info() {
    test_type_reconstruction_internal(
        "test_type_reconstruction_no_size_info",
        PrimitiveReconstructionFlavor::Portable,
        AccessSpecifierReconstructionFlavor::Automatic,
        false,
        true,
        false,
        true,
        false,
        false,
    );
}

#[test]
fn test_type_reconstruction_no_offset_info() {
    test_type_reconstruction_internal(
        "test_type_reconstruction_no_offset_info",
        PrimitiveReconstructionFlavor::Portable,
        AccessSpecifierReconstructionFlavor::Automatic,
        false,
        true,
        true,
        false,
        false,
        false,
    );
}

#[test]
fn test_type_reconstruction_no_comments() {
    test_type_reconstruction_internal(
        "test_type_reconstruction_no_comments",
        PrimitiveReconstructionFlavor::Portable,
        AccessSpecifierReconstructionFlavor::Automatic,
        false,
        true,
        false,
        false,
        false,
        false,
    );
}

fn test_type_reconstruction_internal(
    test_name: &str,
    primitives_flavor: PrimitiveReconstructionFlavor,
    print_access_specifiers: AccessSpecifierReconstructionFlavor,
    reconstruct_dependencies: bool,
    integers_as_hexadecimal: bool,
    print_size_info: bool,
    print_offset_info: bool,
    print_brackets_new_line: bool,
    ignore_std_types: bool,
) {
    let pdb_file = PdbFile::load_from_file(Path::new(TEST_PDB_FILE_PATH)).expect("load test.pdb");
    for (i, test_case_type_name) in TEST_CASES.iter().enumerate() {
        let (reconstructed_type, _) = pdb_file
            .reconstruct_type_by_name(
                test_case_type_name,
                primitives_flavor,
                print_access_specifiers,
                reconstruct_dependencies,
                integers_as_hexadecimal,
                print_size_info,
                print_offset_info,
                print_brackets_new_line,
                ignore_std_types,
            )
            .unwrap_or_else(|_| panic!("reconstruct type: {test_case_type_name}"));

        let snapshot_name = format!("{test_name}-{i}");
        insta::assert_snapshot!(snapshot_name, reconstructed_type);
    }
}
