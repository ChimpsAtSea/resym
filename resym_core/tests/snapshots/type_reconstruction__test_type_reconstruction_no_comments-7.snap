---
source: resym_core/tests/type_reconstruction.rs
assertion_line: 184
expression: reconstructed_type
---
enum resym_test::EnumTest2 : unsigned char {
  kEnumTest2Val1 = 0x0000,
  kEnumTest2Val2 = 0x0001,
  kEnumTest2Val3 = 0x0002,
};
